internal/controllers/machine/machine_controller_noderef.go

```diff
@@ -116,6 +118,17 @@ func (r *Reconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Clust
 		}
 	}
 
+	options := []client.PatchOption{
+		client.FieldOwner("capi-machine"),
+		client.ForceOwnership,
+	}
+	nodeCopy := node.DeepCopy()
+	nodeCopy.Labels = getManagedLabels(machine.Labels)
+	nodePatch := unstructuredNode(nodeCopy)
+	if err := remoteClient.Patch(ctx, nodePatch, client.Apply, options...); err != nil {
```

Comment:
Q: Just wanted to bring up the question if we want to have an envtest based test for this. (could also be in a follow-up PR).

I'm leaning towards no, as it feels like re-testing basic SSA functionality.


On the other side it might be interesting to slightly extend one of our e2e tests:
* adding 1 label per prefix/domain to one of our cluster-templates
* adding code to verify that the node labels have been synced
* maybe also checking that other labels (e.g. added by kubeadm) are still there

Extending the e2e test a bit seems like it would give us a lot of coverage for a small amount of work

@fabriziopandini @ykakarap @enxebre WDYT?

P.S. Definitely fine for me to move any additional tests to a separate follow-up issue

Reply:
+1 to have an e2e. Regardless of using ssa as an impl detail the test has value to validate input/output feature expectations.

Reply:
+1 to including this check in a e2e test. `clusterclass-changes` a probably a good candidate. In addition to the testing that labels are applied we can also test that changing the labels will change the labels on the node too. This will be an easy way to test that dropping labels works as expected.